Fixes for services playback per review 2

MichaelOrlov opened the pull request:
- Relates https://github.com/ros2/rosbag2/pull/1481
## Made the following changes per review:
- Make service_event_ts_lib as private member again
  - Motivation: The shared pointer to the service event type support library shall be a member variable to make sure that the library is loaded during the liveliness of the instance of this class, since we have raw pointers to its inner members.
-  Cleanup in PlayerServiceClient::async_send_request(ser_message)
   - Rewrite raw pointers arithmetic. The assumption that size_t represents the size of the void* may not necessarily be true on all platforms.
   - Use shared pointer with custom deleter for deserialized messages.
   -  The assumption that we can take the first element from the bounded sequence by dereferencing raw pointer to the bounded sequence may not necessarily be true and up to the underlying rmw and transport layer implementation. Use dedicated `request_member.get_function(request_sequence_ptr, 0)` function instead.
   - Add sanity checks `for service_event_members_` in `PlayerServiceClient` constructor.
- Refactoring. Do full deserialization and only once
   - Rationale: We can't rely on the assumption that we can safely partially deserialize service event to the `ServiceEventInfo` structure.

morlov-apexai committed:
Cleanup in PlayerServiceClient::async_send_request(ser_message)

- Rewrite raw pointers arithmetic. Assumption that size_t represent size
of the void* may not necessarily be true on all platforms.
- Use shared pointer with custom deleter for deserialized message.
- Assumption that we can take first element from bounded sequence by
dereferencing raw pointer to the bounded sequence may not be necessarily
be true and up to the underlying rmw and transport layer implementation.
Use dedicated request_member.get_function(request_sequence_ptr, 0)
function instead.
- Add sanity checks `for service_event_members_` in PlayerServiceClient
constructor.

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Make service_event_ts_lib as private member again

- Motivation: The shared pointer to the service event type support
library shall be a member variable to make sure that library loaded
during the liveliness of the instance of this class, since we have a
raw pointers to its inner members.

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Refactoring. Do full deserialization and only once

- Rationale: We can't rely on assumption that we can safely partially
deserialize service event to the ServiceEventInfo structure.

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

MichaelOrlov commented:
@Barry-Xu-2018 @fujitatomoya Here is the draft of my second round changes per review. I haven't done with this PR yet, but you can consider my changes and make some comments. 
I will be on vacation during the next week. However, going to get back to this PR in the middle of the week.

fujitatomoya commented:
@MichaelOrlov thanks we will be reviewing this early in this week.

Barry-Xu-2018 commented on rosbag2_transport/src/rosbag2_transport/player.cpp:
```diff
@@ -1191,17 +1187,61 @@ bool PlayerImpl::publish_message(rosbag2_storage::SerializedBagMessageSharedPtr
+      return false;
+    }
+
+    if (!service_client->is_service_event_include_request_message(service_event)) {
```
It appears to overlook the scenario where both the service and the client enable introspection contents for the same client ID.

Old function is PlayerServiceClient::is_include_request_message().  
Except checking if request is included, this function also includes the following logic
1.  Decide whether the request comes from the service or the client is used for one client ID. (Currently, it is decided by the first message including the request data. e.g. First message with request from client.  We will use message from client to send request, even if message from service also provide request data.) 

2. Consider the case. Service/client switched introspection type from metadata to contents in recorded data. 
I don't consider the case. Service/Client switched introspection type from contents to metadata in record data, which leads that the decision in 1 maybe adjusted.

https://github.com/Barry-Xu-2018/rosbag2/blob/a8312de491a34150ddcb0be4156b7e01ba2ac862/rosbag2_transport/src/rosbag2_transport/player_service_client.cpp#L55-L159

MichaelOrlov replied:
@Barry-Xu-2018 Thanks for finding and pointing out this. It seems I overlook these cases.
From one side I understand that it would be nice to have some autodetection algorithm to be able to decide whether to use a service request from the original client request or from the service request confirmation event message. 
However, this sophisticated logic could have corner cases and would be difficult to understand or explain for final users.
In addition, it seems the original implementation with a "hidden" autodetection algorithm from the `PlayerServiceClient::is_include_request_message()` is not going to work due to the found issue on the RMW layer [[service introspection event does not have unique GID during transaction](https://github.com/ros2/rmw/issues/357#top)
#357](https://github.com/ros2/rmw/issues/357).

I am more inclined to implement a simple logic for a while to allow publish service requests upon a boolean flag by client request or by service confirmation and add this flag to the settings and perhaps to the CLI or env variable. Need to figure out what is better on the current stage.

Also, I think that we need to publish original service events in all other cases to be able to see them in the `topic echo` during playback or other introspection tools.

cc: @fujitatomoya

fujitatomoya replied:
> I am more inclined to implement a simple logic for a while to allow publish service requests upon a boolean flag by client request or by service confirmation and add this flag to the settings and perhaps to the CLI or env variable. Need to figure out what is better on the current stage.

that sounds reasonable to me as well. in default, we could rely the service events on service server side. i think this is easier and suggested way to use, but not always we can do this with 3rd party components.

> Also, I think that we need to publish original service events in all other cases to be able to see them in the topic echo during playback or other introspection tools.

I really do not see the benefit for this, can you share the use case for this?

i would do, `ros2 bag play --service` and enable introspection on server application to see the service server is working okay with `ros2 service echo`. if we publish all the service events on the topic at the same time during playback, that `ros2 service echo` would be really complicated and printing duplicated messages?

or are you suggesting that we would want to publish the service events with topic via specific option?

Barry-Xu-2018 replied:
@MichaelOrlov 

> However, this sophisticated logic could have corner cases and would be difficult to understand or explain for final users.

Yes. Since users are not aware of the internal selection logic, it can lead to confusion about the result under certain scenario.

> I am more inclined to implement a simple logic for a while to allow publish service requests upon a boolean flag by client request or by service confirmation and add this flag to the settings and perhaps to the CLI or env variable. Need to figure out what is better on the current stage.

I also agree with using simpler logic. However, this might result in losing support for complex scenarios, such as some services enabling introspection while others have client introspection enabled. It would require users to uniformly use either service introspection or client introspection.  

I incline to add new option for play.   
e.g.
--service-request-from {service, client}. `service` is default value. 
If there's a need to add automatic selection in the future (This depends on whether there is a demand from the users.), we can add an "auto" option.   

> Also, I think that we need to publish original service events in all other cases to be able to see them in the topic echo during playback or other introspection tools.

Are you saying that users, besides being able to replay service requests, can also replay the service event topic like a normal topic, allowing other introspection tools to analyze recorded the information via service event topic?

If yes, enable this function, is there a new parameter for play (if this parameter exists, play service event topic instead of play service request) ?  About filter parameters (--services, --exclude-services, etc), do they work? Or user should use --topics, --exclude-topics since service event topic is replayed as normal topic.)

Barry-Xu-2018 replied:
>  About filter parameters (--services, --exclude-services, etc), do they work? Or user should use --topics, --exclude-topics since service event topic is replayed as normal topic.)

After checking current code, `--services` and `--exclude-services` should still function normally, affecting the playback of service events.

Barry-Xu-2018 replied:
I try to add a new parameter `--service-request-from {service_introspection,client_introspection}` for play command.   
For the details, refer to https://github.com/Barry-Xu-2018/rosbag2/commit/4d15a21e8e67c7a8dbe2265a3ff5ab1671171ef8

Barry-Xu-2018 replied:
I try to add a new parameter `--publish-service-event` to support playing service event message.  
For the details, refer to https://github.com/Barry-Xu-2018/rosbag2/commit/006bfa9fbc33f23b277f986e3935afdbe5764a36

MichaelOrlov replied:
@Barry-Xu-2018 Thanks for the follow-up and deduction on my proposals. I am sorry I was off for a couple of days. Now, I'm trying to recover.  
As regards adding a new parameter `--service-request-from {service_introspection,client_introspection}` for the play command.
- Yes, the idea is good, and the implementation seems overall correct. I might need to make some minor changes, like nitpicks or review it thoroughly. I will cherry-pick it to my WIP branch and add my edits above as separate commits.

As regards adding a new parameter `--publish-service-event` to support playing service event messages.  
- IMO it will be more consistent to publish services as service events by default since we are saving them this way (just as is) and have an option to `--publish-service-requests`. In this case our other newly added option `--service-request-from {service_introspection,client_introspection}` will be applicable only if the `--publish-service-requests` is enabled.

Barry-Xu-2018 replied:
@MichaelOrlov 

> Yes, the idea is good, and the implementation seems overall correct. I might need to make some minor changes, like nitpicks or review it thoroughly. I will cherry-pick it to my WIP branch and add my edits above as separate commits.

Okay

> IMO it will be more consistent to publish services as service events by default since we are saving them this way (just as is) and have an option to --publish-service-requests. In this case our other newly added option --service-request-from {service_introspection,client_introspection} will be applicable only if the --publish-service-requests is enabled.

Okay.  I start to update https://github.com/Barry-Xu-2018/rosbag2/commit/006bfa9fbc33f23b277f986e3935afdbe5764a36

Barry-Xu-2018 replied:
@MichaelOrlov 

I updated code https://github.com/Barry-Xu-2018/rosbag2/commit/d6db1bb7cb026d99f41061947e3296d0055a7a5a to support `--publish-service-requests`

MichaelOrlov replied:
@Barry-Xu-2018 Sorry didn't have time today to review or cherry-pick your latest updates from https://github.com/Barry-Xu-2018/rosbag2/commit/d6db1bb7cb026d99f41061947e3296d0055a7a5a to support `--publish-service-requests`.

I added API to be able to make tests more deterministic [Add Player::wait_for_sent_service_requests_to_finish() API](https://github.com/Barry-Xu-2018/rosbag2/pull/6/commits/9d98b04e725c7a5e987e930848b9d75d0817d81a).
I would appreciate it if you would rewrite other service-related unit tests in a similar way as I made in the [Make tests play_service_requests_from_service(client) deterministic](https://github.com/Barry-Xu-2018/rosbag2/pull/6/commits/b87c944d5933079c2d61213275b915384d0094b2)

Barry-Xu-2018 replied:
@MichaelOrlov 

Thanks. 

> I would appreciate it if you would rewrite other service-related unit tests in a similar way as I made in the [Make tests play_service_requests_from_service(client) deterministic](https://github.com/Barry-Xu-2018/rosbag2/pull/6/commits/b87c944d5933079c2d61213275b915384d0094b2)

Okay. I will update other service-related unit tests by this way.

MichaelOrlov replied:
@Barry-Xu-2018 Ok I've cherry-picked (https://github.com/Barry-Xu-2018/rosbag2/commit/d6db1bb7cb026d99f41061947e3296d0055a7a5a to support `--publish-service-requests`) and made some fixes above it. See my new commits on this branch.

Barry-Xu-2018 replied:
@MichaelOrlov 

> I will update other service-related unit tests by this way.

I updated test code based on latest code. Please refer to https://github.com/Barry-Xu-2018/rosbag2/commit/e09632321a0cc4433a193e4c54559e6da6ff0657.

MichaelOrlov replied:
@Barry-Xu-2018 Thanks for updating tests in the https://github.com/Barry-Xu-2018/rosbag2/commit/e09632321a0cc4433a193e4c54559e6da6ff0657 I will review your changes at my Thursday morning or first half of the day.
Meanwhile could you please review my changes per this PR one more time?

MichaelOrlov replied:
@Barry-Xu-2018 I've cherry-picked your changes in tests from https://github.com/Barry-Xu-2018/rosbag2/commit/e09632321a0cc4433a193e4c54559e6da6ff0657. Honestly, didn't have time to review thoroughly, but this shouldn't be a problem. 
If you don't have any more comments/requests for changes for This PR let's move forward and merge rhi PR to your branch. 
I think we almost good to go. Will need to rebase and run CI after merge.

MichaelOrlov replied:
I have some nitpick in mind to fix - but we can do it after code freeze

Barry-Xu-2018 commented:
@MichaelOrlov 

Thank you for your revision. I have no further comments aside from the previous one.

fujitatomoya commented on rosbag2_cpp/src/rosbag2_cpp/service_utils.cpp:
```diff
@@ -132,6 +132,8 @@ size_t get_serialization_size_for_service_metadata_event()
     static_cast<const rosidl_typesupport_introspection_cpp::MessageMembers *>(
     type_support_handle->data);
 
+  // TODO(morlov): We shall not rely on this arithmetic!!! It is up to the serialization
+  //  implementation
```
trying make sure it can find either of us by github handle names.

```suggestion
  // TODO(@MichaelOrlov @fujitatomoya @Barry-Xu-2018 )
  // We shall not rely on this arithmetic!!! It is up to the serialization implementation.
```

MichaelOrlov replied:
@fujitatomoya I am going to delete this code and put a comment on why it is not acceptable to go with it.
Sorry, I didn't have time to add a comment about it in the review.

fujitatomoya reviewed (commented):
overall looks good to me.

Barry-Xu-2018 committed:
Specify service request from which introspection message and fix uncrustify errors

Signed-off-by: Barry Xu <barry.xu@sony.com>

Barry-Xu-2018 committed:
Add a new CLI parameter "--publish-service-requests" for Player

Also added a new option publish_service_requests to the PlayOptions.
Note: By default rosbag2 player will publish service events only.

Signed-off-by: Barry Xu <barry.xu@sony.com>
Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Revert uncrustify changes from previous commit.

- Rationale: We are moving to the new version of the uncrustify in
rolling and still haven't done it yet fully for the baseline.
The discrepancy in style for other untouched files like logging.hpp
shall be addressed in a separate PR.

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

MichaelOrlov commented on ros2bag/ros2bag/verb/play.py:
```diff
@@ -152,6 +153,11 @@ def add_arguments(self, parser, cli_name):  # noqa: D102
                  'By default, if loaned message can be used, messages are published as loaned '
                  'message. It can help to reduce the number of data copies, so there is a greater '
                  'benefit for sending big data.')
+        parser.add_argument(
+            '--service-request-from', default='service_introspection',
+            choices=['service_introspection', 'client_introspection'],
+            help='Determine the source of the service request to be replayed. '
+                 'By default, the service request is from recorded service introspection message.')
```
@Barry-Xu-2018 I am considering to rename the `--service-request-from` parameter and `service_request_from` play option to the "--service-request**s**-**source**" and "play_options.service_request**s**_**source**" respectively.
Also use short names for enum values without `INTROSPECTION` postfix. 
i.e., `ServiceRequestsSource.SERVICE` and `ServiceRequestsSource.CLIENT`.
IMO '"from"'  postfix usually is not used for parameters or variables, the `source` is a better fit for that purpose.

If you agree I can make changes on my branch as a separate commit.

Barry-Xu-2018 replied:
@MichaelOrlov 

I agree with your suggestions. Thanks！

morlov-apexai committed:
Rename service_request_from to the service_requests_source

- Also rename enum class ServiceRequestFrom to the ServiceRequestsSource

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Add Player::wait_for_sent_service_requests_to_finish() API

- We need this API to be able to write deterministic and non-flaky tests

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Mitigate potential issues related to the operations reordering on ARM

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Make tests play_service_requests_from_service(client) deterministic

- Get rid of timeout inside tests and use newly added
player->wait_for_sent_service_requests_to_finish(service_name) API.

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Misc findings and improvements 1

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Rename get_services_clients() to the get_service_clients()

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Fix an issue on filtering topic when prepare publishers

Co-authored-by: Barry Xu <barry.xu@sony.com>
Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Cleanup in play_without_publish_service_requests

- Long story short: Make it deterministic and run fast.

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

Barry-Xu-2018 committed:
Update test codes

Signed-off-by: Barry Xu <barry.xu@sony.com>

morlov-apexai committed:
Wrap code which can throw with try-catch in the publish_message(..)

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Delete some part of the code which became absolute and shall not be used

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

Barry-Xu-2018 commented on rosbag2_transport/src/rosbag2_transport/player.cpp:
```diff
@@ -787,6 +795,29 @@ void PlayerImpl::delete_on_play_message_callback(const callback_handle_t & handl
+    auto service_event_name = rosbag2_cpp::service_name_to_service_event_topic_name(service_name);
+    auto client_iter = service_clients_.find(service_event_name);
+    if (client_iter != service_clients_.end()) {
+      is_requests_complete = client_iter->second->wait_for_sent_requests_to_finish(timeout);
```
Finally, it also will call PlayerServiceClientManager::wait_for_all_futures().   
So this operation does not wait for the future of the client of the specified service to complete, but waits for all futures to complete.

MichaelOrlov replied:
@Barry-Xu-2018 In this particular case it will wait for all futures for specific client.
However if `service_name` is empty it will wait for all futures from all clients, - does it make sense?

Barry-Xu-2018 replied:
@MichaelOrlov 

> However if service_name is empty it will wait for all futures from all clients, - does it make sense?

Yes. It makes sense. 

> In this particular case it will wait for all futures for specific client.

Although the code retrieves the corresponding client based on the service name, `client->wait_for_sent_requests_to_finish()` actually calls `PlayerServiceClientManager::wait_for_all_futures()`. This function does not wait for all futures of a specific client; instead, it waits for all futures of all registered clients.

Barry-Xu-2018 commented on rosbag2_cpp/src/rosbag2_cpp/service_utils.cpp:
```diff
@@ -33,54 +33,54 @@ bool is_service_event_topic(const std::string & topic, const std::string & topic
-  }
+    if (topic_type.length() <= std::strlen(service_event_topic_type_postfix)) {
+      return false;
+    }
```
This code is unnecessary since this code is else part of line 46.

Barry-Xu-2018 commented on rosbag2_cpp/src/rosbag2_cpp/service_utils.cpp:
```diff
@@ -33,54 +33,54 @@ bool is_service_event_topic(const std::string & topic, const std::string & topic
+    }
+
+    service_name = topic_name.substr(
+      0, topic_name.length() - strlen(RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX));
```
Perhaps we could simplify these codes.

```c++
} else {
  if (topic_name.substr(
        topic_name.length() -
        strlen(RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX)) ==
      RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX)
    {
        service_name = topic_name.substr(
           0, topic_name.length() - strlen(RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX));
    } 
   return service_name;
}
```

Barry-Xu-2018 commented on rosbag2_transport/src/rosbag2_transport/player.cpp:
```diff
@@ -1221,10 +1221,8 @@ bool PlayerImpl::publish_message(rosbag2_storage::SerializedBagMessageSharedPtr
   // Try to publish message as service request
   auto client_iter = service_clients_.find(message->topic_name);
-  if (client_iter != service_clients_.end()) {
+  if (play_options_.publish_service_requests && client_iter != service_clients_.end()) {
```
I think `play_options_.publish_service_requests` is unnecessary.   
In the process of PlayerImpl::prepare_publishers(), `play_options_.publish_service_requests` can make sure service event topic is in publishers_.   
So if play_options_.publish_service_requests is not set, the service event topic is already handled in the code for publisher_.

MichaelOrlov replied:
@Barry-Xu-2018 I would keep it for defensive programming or if in the future one will decide to change the value of the `play_options_.publish_service_requests` in runtime.

Barry-Xu-2018 commented:
@MichaelOrlov 

I have a few minor comments. 

> If you don't have any more comments/requests for changes for This PR let's move forward and merge rhi PR to your branch.

I will merge this PR. And above minor comments, I will continue to do in https://github.com/ros2/rosbag2/pull/1481
Thank you.
